lib/Parse/Lexer.cpp

```diff
@@ -278,8 +278,14 @@ void Lexer::formToken(tok Kind, const char *TokStart, bool MultilineString) {
     Kind = tok::eof;
   }
   unsigned CommentLength = 0;
-  if (RetainComments == CommentRetentionMode::AttachToNextToken && SeenComment)
-    CommentLength = TokStart - LastCommentBlockStart;
+  if (RetainComments == CommentRetentionMode::AttachToNextToken) {
+    auto FirstCommentIter = std::find_if(
+        LeadingTrivia.begin(), LeadingTrivia.end(),
+        [](const TriviaPiece &Piece) { return Piece.isComment(); });
+    for (auto Iter = FirstCommentIter; Iter != LeadingTrivia.end(); Iter++) {
```

Comment:
Small code style request:
```cpp
    auto Iter = ...;
    for (auto End = LeadingTrivia.end(); Iter != End; ++Iter) {
```

Also, try using [`llvm::find_if`](http://llvm.org/doxygen/STLExtras_8h_source.html) which accept a range instead of iterators.

Reply:
ok, I will do later.